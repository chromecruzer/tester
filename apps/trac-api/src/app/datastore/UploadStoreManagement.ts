import path from "path";
import FileStore from "./FileStore";
import DataStoreUtilities from "./DataStoreUtilities";
const storagePathFn = uploadConfig => path.join(...uploadConfig.root, 'session-data');
export default class UploadStoreManagement {
  private results: { [id: string]: FileStore } = {};
  private storagePath: string;
  constructor(uploadConfig) {
    this.storagePath = storagePathFn(uploadConfig);
  }
  static async createStorageDir(uploadConfig, testDsu = null) {
    const dsu = testDsu || new DataStoreUtilities();
    return dsu.mkdir(storagePathFn(uploadConfig));
  }
  public async create(session, json) {
    const store = new FileStore(this.storagePath, session);
    // console.log(`create storage path ${this.storagePath} with ${session}`)
    return store.set(json).then(() => {
      this.results[session] = store;
    });
  }
  private async recreate(session) {
    const store = new FileStore(this.storagePath, session);
    // console.log(`recreate storage path ${this.storagePath} with ${session}`)
    const json = await store ? store.get() : null;
    this.results[session] = store;
    return json;
  }

  public async retrieve(session) {
    // console.log(`retrieving ${session}`)
    if(this.results[session]) {
      return this.results[session].get();
    }
    return this.recreate(session);
  }
  public async removeDataStore(session) {
    return this.results[session].remove().then(() => {
      delete this.results[session];
    });
  }

}
