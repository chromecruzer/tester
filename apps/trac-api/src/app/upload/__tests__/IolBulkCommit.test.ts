import IolBulkCommit from "../IolBulkCommit";
import UploadStoreManagement from "../../datastore/UploadStoreManagement";
import util from "util";

const dump = (obj, depth = null) => util.inspect(obj, {depth, colors: true});

jest.mock('../../datastore/UploadStoreManagement');
const comparisons = {
  adds: [
    {
      uuid: 'uuid1',
      warehouse: 'CLR34',
      location_code: 'LOC1',
      name_code: 'nameCode1',
      address_phone: '123 White House Lane 888-112-1234',
      item: 'item1',
      description: 'I describe an item1',
      lot: 'lot1223',
      quantity: 1,
      purchase_order: 'purchased for dimplewig as item1',
      received_date: new Date('10/10/2021'),
      expire_date: new Date('10/10/2024'),
      shipped_date: new Date('10/07/2021'),
    },
    {
      uuid: 'uuid2',
      warehouse: 'CLR34',
      location_code: 'LOC1',
      name_code: 'nameCode2',
      address_phone: '123 Black House Lane 888-112-1234',
      item: 'item2',
      description: 'I describe an item2',
      lot: 'lot1225',
      quantity: 1,
      purchase_order: 'purchased for dimplewig as item1',
      received_date: new Date('12/10/2021'),
      expire_date: new Date('01/10/2024'),
      shipped_date: new Date('12/07/2021'),
    }
  ],
  removedUuids: ['uuid3', 'uuid4'],
  modifies: [
    {
      originalUuid: 'uuid5',
      new: {
        uuid: 'uuid5',
        warehouse: 'CLR34',
        location_code: 'LOC1',
        name_code: 'nameCode5',
        address_phone: '123 Vatican Alley 888-112-1234',
        item: 'item5',
        description: 'I describe an item5',
        lot: 'lot1445',
        quantity: 1,
        purchase_order: 'purchased for the congregation',
        received_date: new Date('01/10/2021'),
        expire_date: new Date('02/10/2024'),
        shipped_date: new Date('01/07/2021'),
      },
    },
    {
      originalUuid: 'uuid6',
      new: {
        uuid: 'uuid6',
        warehouse: 'CLR34',
        location_code: 'LOC1',
        name_code: 'nameCode6',
        address_phone: '123 SoCal Alley 888-112-1234',
        item: 'item7',
        description: 'I describe an item7',
        lot: 'lot1645',
        quantity: 1,
        purchase_order: 'purchased for the public',
        received_date: new Date('02/10/2021'),
        expire_date: new Date('03/10/2024'),
        shipped_date: new Date('02/07/2021'),
      },
    },
  ],
}
describe('IolBulkCommits', () => {
  describe('#commit', () => {
    //TODO: Some kind of bug ignores the mocked version of fill.
    xit('should call bulk fill and delete with the proper arguments', () => {
      const dataStore = new UploadStoreManagement('config');
      const mockRetrieve = jest.spyOn(dataStore, 'retrieve').mockReturnValue(Promise.resolve(comparisons));
      const bulkCommits = new IolBulkCommit(dataStore, 'schema name', 'table name');
      const mockDelete = jest.spyOn(bulkCommits.bulkForNew, 'delete').mockReturnValue(Promise.resolve());
      const mockFill = jest.spyOn(bulkCommits.bulkForNew, 'fill');
      console.log(`bulk fill ${dump(bulkCommits.bulkForNew.fill)}`);
      return bulkCommits.commit(comparisons, 'client').then(() => {
        // console.log('evaluating commit call');
        expect(mockDelete.mock.calls[0][0]).toEqual(['uuid3', 'uuid4', 'uuid5', 'uuid6']);
        expect(mockDelete.mock.calls[0][1]).toEqual('client');
        expect(mockFill.mock.calls[0][1]).toEqual('client');
        expect(mockFill.mock.calls[0][0]).toMatchSnapshot();
      });
    });
  });
});
