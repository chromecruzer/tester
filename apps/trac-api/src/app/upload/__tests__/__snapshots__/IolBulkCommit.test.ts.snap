// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IolBulkCommits #commit should call bulk fill and delete with the proper arguments 1`] = `
Array [
  Array [
    "uuid1",
    "CLR34",
    "LOC1",
    "nameCode1",
    "123 White House Lane 888-112-1234",
    "item1",
    "I describe an item1",
    "lot1223",
    1,
    "purchased for dimplewig as item1",
    2021-10-10T05:00:00.000Z,
    2024-10-10T05:00:00.000Z,
    2021-10-07T05:00:00.000Z,
  ],
  Array [
    "uuid2",
    "CLR34",
    "LOC1",
    "nameCode2",
    "123 Black House Lane 888-112-1234",
    "item2",
    "I describe an item2",
    "lot1225",
    1,
    "purchased for dimplewig as item1",
    2021-12-10T06:00:00.000Z,
    2024-01-10T06:00:00.000Z,
    2021-12-07T06:00:00.000Z,
  ],
  Array [
    "uuid5",
    "CLR34",
    "LOC1",
    "nameCode5",
    "123 Vatican Alley 888-112-1234",
    "item5",
    "I describe an item5",
    "lot1445",
    1,
    "purchased for the congregation",
    2021-01-10T06:00:00.000Z,
    2024-02-10T06:00:00.000Z,
    2021-01-07T06:00:00.000Z,
  ],
  Array [
    "uuid6",
    "CLR34",
    "LOC1",
    "nameCode6",
    "123 SoCal Alley 888-112-1234",
    "item7",
    "I describe an item7",
    "lot1645",
    1,
    "purchased for the public",
    2021-02-10T06:00:00.000Z,
    2024-03-10T06:00:00.000Z,
    2021-02-07T06:00:00.000Z,
  ],
]
`;
